src/main/scala/com/microsoft/hyperspace/actions/RefreshDeleteAction.scala

```diff
@@ -0,0 +1,104 @@
+import com.microsoft.hyperspace.index.DataFrameWriterExtensions.Bucketizer
+import com.microsoft.hyperspace.telemetry.{AppInfo, HyperspaceEvent, RefreshDeleteActionEvent}
+import com.microsoft.hyperspace.util.ResolverUtils
+
+/**
+ * Refresh index by removing index entries from any deleted source data file.
+ * Note this Refresh Action only fixes an index w.r.t deleted source data files
+ * and does not consider new source data files (if any).
+ * If some original source data file(s) are removed between previous version of index and now,
+ * this Refresh Action updates the index as follows:
+ * 1. Deleted source data files are identified;
+ * 2. Index records' lineage is leveraged to remove any index entry coming
+ *    from those deleted source data files.
+ *
+ * @param spark SparkSession
+ * @param logManager Index LogManager for index being refreshed
+ * @param dataManager Index DataManager for index being refreshed
+ */
+class RefreshDeleteAction(
+    spark: SparkSession,
+    logManager: IndexLogManager,
+    dataManager: IndexDataManager)
+    extends RefreshActionBase(spark, logManager, dataManager) with Logging {
+
+  final override protected def event(appInfo: AppInfo, message: String): HyperspaceEvent = {
+    RefreshDeleteActionEvent(appInfo, logEntry.asInstanceOf[IndexLogEntry], message)
+  }
+
+  /**
+   * For an index with lineage, find all the source data files which have been deleted,
+   * and use index records' lineage to mark and remove index entries which belong to
+   * deleted source data files as those entries are no longer valid.
+   */
+  final override def op(): Unit = {
+    val indexDF = spark.read.parquet(previousIndexLogEntry.content.files.map(_.toString): _*)
+
+    ResolverUtils
+      .resolve(spark, IndexConstants.DATA_FILE_NAME_COLUMN, indexDF.schema.fieldNames) match {
+      case Some(_) =>
+        val deletedFiles = getDeletedFiles
+        logInfo("Refresh index is updating index by removing index entries " +
+          s"from ${deletedFiles.length} deleted source data files.")
+
+        val refreshDF =
+          indexDF.filter(
+            !col(s"${IndexConstants.DATA_FILE_NAME_COLUMN}").isin(deletedFiles: _*))
+
+        refreshDF.write.saveWithBuckets(
+          refreshDF,
+          indexDataPath.toString,
+          previousIndexLogEntry.numBuckets,
+          indexConfig.indexedColumns)
+
+      case None =>
+        throw HyperspaceException(
+          s"Index refresh (to handle deleted source data) is" +
+            " only supported on an index with lineage.")
+    }
+  }
+
+  /**
+   * Compare list of source data files from previous IndexLogEntry to list of
+   * current source data files and identify deleted source data files.
+   *
+   * @return list of full paths of deleted source data files.
+   */
+  private def getDeletedFiles: Seq[String] = {
+    val rels = previousIndexLogEntry.relations
+    val originalFiles = rels.head.data.properties.content.files.map(_.toString)
```

Comment:
Could you check the file metadata (size / modification time ) as well?

Reply:
@sezruby Here we are looking for "deleted" files and file name suffices for that check. If a file is renamed we still mark it as deleted, however I am not sure if modifying contents of an existing file is a valid scenario. Can you explain a bit what exactly you are suggesting? thnx

Reply:
Good point. I think the user can "overwrite" the files.

I think if we detect a mismatch of metadata, we shouldn't perform the action, but suggest to perform full refresh.

Reply:
@imback82 Does that mean for existing files (those which are present both in `originalFiles` and `currentFiles`) we should do a full metadata comparison here and if there is a mismatch we abort on-going refresh action?

Reply:
+1 It might be better to just do a full metadata comparison and abort (with a suggestion). It is becoming increasingly clear that we should be on the safe side. :)

Reply:
I think we can handle the modified file as both deleted and appended file. Aborting DeleteRefresh in this PR for now, but later with append impl. we could delete & append refresh the file properly.

Reply:
FileInfo check added along with a new test case under `RefreshIndexTests.scala` to validate it.

Reply:
@apoorvedave1 as FYI